builtin/providers/aws/data_source_aws_sns.go

```diff
@@ -0,0 +1,69 @@
+package aws
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceAwsSnsTopic() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsSnsTopicsRead,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceAwsSnsTopicsRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).snsconn
+	params := &sns.ListTopicsInput{
+		NextToken: aws.String("nextToken"),
+	}
+
+	target := d.Get("name")
+
+	validNamePattern := "^[A-Za-z0-9_-]+$"
```

Comment:
I think this should be moved to a ValidateFunc as part of the schema - that way it can be caught earlier